core/shared/src/main/scala/eu/joaocosta/minart/graphics/Kernel.scala

```diff
@@ -0,0 +1,218 @@
+package eu.joaocosta.minart.graphics
+
+/** A kernel used for convolution.
+  * It simplifies the usage of Plane#coflatMap in a performant way.
+  */
+sealed trait Kernel {
+
+  def width: Int
+
+  def height: Int
+
+  /** Convolution function to use.
+    *
+    * Use as `plane.coflatMap(kernel)`
+    */
+  def apply(getPixel: (Int, Int) => Color): Color
+}
+
+object Kernel {
+
+  def apply(matrix: Seq[Seq[Int]] = Seq(Seq(1)), normalization: Int = 1, constant: Int = 0): SingleKernel =
+    new SingleKernel(IArray.from(matrix.iterator.map(IArray.from)), normalization, constant: Int)
+
+  /** Identity kernel, does nothing.
+    */
+  val identity: SingleKernel = SingleKernel()
+
+  /** Average blur kernel. Averages the pixels in a window.
+    *
+    *  Both the width and height must be odd.
+    */
+  def averageBlur(width: Int, height: Int): SingleKernel = {
+    val line = IArray.fill(width)(1)
+    SingleKernel(IArray.fill(height)(line), width * height)
+  }
+
+  /** Horizontal Sobel operator. Returns the horizontal derivative of the image.
+    * Since colors cannot have a negative value, a derivative of 0 corresponds to 127.
+    */
+  val horizontalSobel: SingleKernel =
+    SingleKernel(IArray(IArray(-1, 0, 1), IArray(-2, 0, 2), IArray(-1, 0, 1)), 8, 127)
+
+  /** Vertical Sobel operator. Returns the vertical derivative of the image.
+    * Since colors cannot have a negative value, a derivative of 0 corresponds to 127.
+    */
+  val verticalSobel: SingleKernel =
+    SingleKernel(IArray(IArray(-1, -2, -1), IArray(0, 0, 0), IArray(1, 2, 1)), 8, 127)
+
+  /** Convolution kernel using a single matrix., where all color chanels are handled the same way.
+    * The alpha channel is discarded (forced to 255).
+    *
+    * The center of the matrix maps to the target position. As such, all dimensions must be odd.
+    *
+    * For performance reasons, all entries are integers, but are then divided by normalization
+    * constant.
+    *
+    * A constant value can also be added. This value is added only after the normalization.
+    */
+  final case class SingleKernel(
+      matrix: IArray[IArray[Int]] = IArray(IArray(1)),
```

Comment:
Note for my future self: I had some performance issues in Scala Native by using `ArraySeq` here.

Maybe it's possible to use `ArraySeq` without the performance impact by accessing the `unsafeArray`.